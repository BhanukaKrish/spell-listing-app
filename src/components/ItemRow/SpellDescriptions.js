import React from "react";
import style from "./style.module.css";

const SpellDescriptions = ({ spellData }) => {
  return (
    <div className={style.expandBox}>
      {/* Spell description  */}
      <h2>Description</h2>
      {spellData &&
        spellData.desc?.map((item, index) => <p key={index}>{item}</p>)}

      {/* Spell Higher Level  */}
      {spellData &&
        spellData.higher_level &&
        spellData.higher_level.length > 0 && (
          <>
            <h2>Higher Level</h2>
            {spellData.higher_level?.map((item, index) => (
              <p key={index}>{item}</p>
            ))}
          </>
        )}
      {/* Spell key tags  */}
      {spellData && (
        <div className={style.tagRow}>
          {/* Range */}
          <div className={style.tagBox}>Range : {spellData?.range}</div>

          {/* Material */}
          <div className={style.tagBox}>Material : {spellData?.material}</div>

          {/* Duration */}
          <div className={style.tagBox}>Duration : {spellData?.duration}</div>

          {/* Casting Time */}
          <div className={style.tagBox}>
            Casting Time : {spellData?.casting_time}
          </div>

          {/* Level */}
          <div className={style.tagBox}>Level : {spellData?.level}</div>

          {/* Attack Type */}
          <div className={style.tagBox}>
            Attack Type : {spellData?.attack_type}
          </div>

          {/* Spell School */}
          <div className={style.tagBox}>School : {spellData?.school?.name}</div>

          {/* Spell Classes */}
          <div className={style.tagBox}>
            Classes :
            {spellData?.classes?.map((item, index) => (
              <p key={index}>{item.name}</p>
            ))}
          </div>

          {/* Spell Subclasses */}
          <div className={style.tagBox}>
            Sub Classes :
            {spellData?.subclasses?.map((item, index) => (
              <p key={index}>{item.name}</p>
            ))}
          </div>
        </div>
      )}
    </div>
  );
};

export default SpellDescriptions;
