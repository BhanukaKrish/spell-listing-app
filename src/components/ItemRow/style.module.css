.box{
    margin-top: 5px;
    width: 100%;
    /* height: 100vh; */
    background-color: var(--color-white);
    border-radius: 10px;
    transition: var(--transition);
    overflow: auto;
    padding: 13px 20px;
    color: var(--color-text);
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
}

.box > p{
    font-size: var(--font-size-lg);
    font-weight: 300;
}

.box:hover{
    background-color: var(--color-secondary);
    box-shadow: var(--box-shadow-base);
    color: var(--color-white)
}

.box_btn{
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
    gap: 20px;
}

.expandBtn{
    cursor: pointer;
    margin-top: 5px;
    font-size: 20px;
}

.faveBtn{
    cursor: pointer;
    font-size: 20px;
}

.expandBox{
    width: 100%;
    background-color: whitesmoke;
    border-radius: 10px;
    transition: var(--transition);
    padding: 13px 20px;
    color: var(--color-text);
    display: flex;
    flex-direction: column;
}

.expandBox > h2{
    font-size: var(--font-size-lg);
    font-weight: 600;
    margin-top: 10px;
}

.tagRow{
    margin-top: 20px;
}

.tagBox{
    width: fit-content;
    background-color: var(--color-secondary);
    padding: 5px 10px;
    color: var(--color-white);
    border-radius: 20px;
    margin-top: 5px;
    font-size: var(--font-size-base);
    display: flex;
    align-items: center;
    gap: 10px;
}