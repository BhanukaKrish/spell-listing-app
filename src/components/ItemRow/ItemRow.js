import React, { useState } from "react";
import style from "./style.module.css";
import { IoIosArrowDown } from "react-icons/io";
import { IoIosArrowUp } from "react-icons/io";
import { MdFavoriteBorder } from "react-icons/md";
import { MdFavorite } from "react-icons/md";
import useSpellData from "../../hooks/useSpellData";
import SpellDescriptions from "./SpellDescriptions";
import { useFaveStore } from "../../store";

function ItemRow({ item }) {
  // * Get the favourite spells from the store.
  const setSpellFave = useFaveStore((state) => state.setSpellFave);
  const spellFave = useFaveStore((state) => state.spellFave);

  const [show, setShow] = useState(false);
  const [selectedSpell, setSelectedSpell] = useState(null);
  const getSpellData = useSpellData(selectedSpell);

  /**
   * Toggle the show/hide state of the spell description.
   * @param {string} spell - spell index value.
   * @param {string} flag - show/hide flag.
   */
  const handleShow = (spell, flag) => {
    setSelectedSpell(spell);
    setShow(flag);
  };

  /**
   * Toggle the favourite spell.
   * @param {string} spell - The spell object.
   */
  const handleFavorite = (spell) => {
    const fave = [...spellFave];
    if (fave.filter((item) => item.index === spell.index).length > 0) {
      fave.splice(fave.indexOf(spell), 1);
    } else {
      fave.push(spell);
    }
    setSpellFave(fave);
  };

  return (
    <>
      <div className={style.box}>
        <p>{item.name}</p>
        <div className={style.box_btn}>
          {spellFave.find((faveItem) => faveItem.index === item.index) ? (
            <div className={style.faveBtn} onClick={() => handleFavorite(item)}>
              <MdFavorite />
            </div>
          ) : (
            <div className={style.faveBtn} onClick={() => handleFavorite(item)}>
              <MdFavoriteBorder />
            </div>
          )}
          
          {show ? (
            <div
              className={style.expandBtn}
              onClick={() => handleShow(null, false)}
            >
              <IoIosArrowUp />
            </div>
          ) : (
            <div
              className={style.expandBtn}
              onClick={() => handleShow(item.index, true)}
            >
              <IoIosArrowDown />
            </div>
          )}
        </div>
      </div>
      {show && <SpellDescriptions spellData={getSpellData} />}
    </>
  );
}

export default ItemRow;
