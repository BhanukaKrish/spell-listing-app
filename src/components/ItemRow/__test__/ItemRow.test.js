import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import ItemRow from "../ItemRow";
import { MemoryRouter } from "react-router-dom";

describe("Test the ItemRow Component", () => {
    test("check the props pass correctly", () => {
        render(<ItemRow item={{ name: "Test" }} />, { wrapper: MemoryRouter });
        const name = screen.getByText("Test");
        expect(name).toBeInTheDocument();
    });

    test("Snapshot of ItemRow", () => {
        const { container } = render(<ItemRow item={{ name: "Test" }} />, { wrapper: MemoryRouter });
        expect(container).toMatchSnapshot();
    });

});
