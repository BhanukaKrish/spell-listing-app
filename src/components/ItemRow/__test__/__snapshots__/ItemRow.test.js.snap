// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the ItemRow Component Snapshot of ItemRow 1`] = `
<div>
  <div
    class="box"
  >
    <p>
      Test
    </p>
    <div
      class="box_btn"
    >
      <div
        class="faveBtn"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
          />
        </svg>
      </div>
      <div
        class="expandBtn"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
